MON-4386,MON-4387: `telemetry` Collection Profile

rexagod opened the pull request:
<!--
    Don't forget about CHANGELOG if this affects the end user!

    Changelog entry format:
    - [#<PR-id>](<PR-URL>) Monitoring <Component> ...

    <PR-id> Id of your pull request.
    <PR-URL> URL of your PR
    <Component> Component affected by your changes such as deps bump, alerts changes and any user facing changes.

    Example:
    - [#741](https://github.com/openshift/cluster-monitoring-operator/pull/741) Bump thanos components to v0.11.0 release
-->

* [x] I added CHANGELOG entry for this change.
* [ ] No user facing changes, so no entry in CHANGELOG was needed.

***

#### Both of these changes need to be merged together. Additionally, [MON-4788](https://issues.redhat.com/browse/MON-4388) also needs to be merged with these (once it's pushed).

***

Marked as draft for now as this needs all non-in-cluster-monitoring-stack recording rules ported to monitors as well.

Not 100% sure right now without investigating more on this, but I believe we can keep the existing machinery, i.e., the whitelisted machinery in CMO and everything from that point on as is, as that would serve as (a) a safeguard to make sure we pipe out the same rules and (b) a check to make sure the metrics (rules) being collected match the rules being sent out. But this will be worked on later once we start adding tests for this in origin (after the aforementioned porting effort is done).

simonpasquier commented on jsonnet/utils/generate-service-monitors.libsonnet:
```diff
@@ -2,13 +2,16 @@
-      name+: '-minimal',
-      labels+: minimalLabel,
+      name+: '-' + label['monitoring.openshift.io/collection-profile'],
+      labels+: label,
```
(nit)
```suggestion
  local run(sm, metrics, profile) = sm {
metadata+: {
      name+: '-' + profile,
      labels+: {
          'monitoring.openshift.io/collection-profile': profile,
       },
```

simonpasquier commented on pkg/tasks/telemeter.go:
```diff
@@ -187,14 +187,16 @@ func (t *TelemeterClientTask) create(ctx context.Context) error {
+	for _, sm := range sms {
+		err = t.client.CreateOrUpdateServiceMonitor(ctx, sm)
+		if err != nil {
+			return fmt.Errorf("reconciling Telemeter client ServiceMonitor failed: %w", err)
```
(nit) should we print the name of the resource?

simonpasquier commented on pkg/manifests/manifests.go:
```diff
@@ -430,10 +438,22 @@ func (f *Factory) AlertmanagerUserWorkloadClusterRole() (*rbacv1.ClusterRole, er
+func (f *Factory) AlertmanagerServiceMonitors() ([]*monv1.ServiceMonitor, error) {
+	return serviceMonitors(
+		f.AlertmanagerServiceMonitor,
+		nil,
```
is it expected that we have a nil servicemonitor for the minimal profile?

rexagod replied:
I'd say so. Currently AM has no metrics that are applied under `minimal` circumstances. However, certain telemetry rules do rely on some of it's metrics so we have one for that profile.

rexagod replied:
Right. It seems we've identified a gap in the `minimal` implementation.

The `minimal` implementation had and has no way to update itself based on the telemetry metrics being added over time. I'll make some jsonnet changes here to merge all `telemetry` profile metrics into `minimal`, so the former is truly a subset of the latter, and we do not miss any metrics as they get added.

simonpasquier commented on jsonnet/components/cluster-monitoring-operator.libsonnet:
```diff
@@ -159,6 +160,12 @@ function(params) {
 
+  telemetryServiceMonitor: generateServiceMonitor.telemetry(
+    self.serviceMonitor, std.join('|', [
+      'cluster_monitoring_operator_collection_profile',
```
how do we ensure that the we keep the list up-to-date?

rexagod replied:
The following rules from the whitelist are currently reliant on us.

```
  '{__name__="cluster:cpu_usage_cores:sum"}'
  '{__name__="ALERTS",alertstate="firing",severity=~"critical|warning|info|none"}'
  '{__name__="cluster:alertmanager_integrations:max"}'
  '{__name__="cluster:capacity_cpu_cores:sum"}'
  '{__name__="cluster:capacity_memory_bytes:sum"}'
  '{__name__="cluster:memory_usage_bytes:sum"}'
  '{__name__="cluster:node_instance_type_count:sum"}'
  '{__name__="cluster:telemetry_selected_series:count"}'
  '{__name__="cluster:virt_platform_nodes:sum"}'
  '{__name__="count:up0"}'
  '{__name__="count:up1"}'
  '{__name__="monitoring:container_memory_working_set_bytes:sum"}'
  '{__name__="namespace_job:scrape_samples_post_metric_relabeling:topk3"}'
  '{__name__="namespace_job:scrape_series_added:topk3_sum1h"}'
  '{__name__="node_role_os_version_machine:cpu_capacity_cores:sum"}'
  '{__name__="node_role_os_version_machine:cpu_capacity_sockets:sum"}'
  '{__name__="openshift:cpu_usage_cores:sum"}'
  '{__name__="openshift:memory_usage_bytes:sum"}'
  '{__name__="openshift:prometheus_tsdb_head_samples_appended_total:sum"}'
  '{__name__="openshift:prometheus_tsdb_head_series:sum"}'
  '{__name__="profile:cluster_monitoring_operator_collection_profile:max"}'
  '{__name__="vendor_model:node_accelerator_cards:sum",vendor=~"NVIDIA|AMD|GAUDI|INTEL|QUALCOMM|Marvell|Mellanox"}'
  '{__name__="workload:cpu_usage_cores:sum"}'
  '{__name__="workload:memory_usage_bytes:sum"}'
  '{__name__=~"cluster:usage:.*"}'
```

I'm thinking of marking these rules in the telemetry ConfigMap (that we use as the source of truth for the test introduced in https://github.com/openshift/origin/pull/30643). We'll extract and query these to ensure they are working. If not, it would likely mean that the SMs are outdated and not updated with relevant metrics.

We'll document this in the handbook so teams can mark rules they introduce when any of them rely on one of our components, as well as update the `jsonnet` code. However, this does make reviewing a bit more involved for telemetry rule addition PRs.

rexagod replied:
The variance-driven test that I have in the aforementioned PR is more of a good measure. I believe the approach above will be more robust.

simonpasquier reviewed (commented):
Good work! I'm still worried that the metrics list get out of date compared to what telemetry expects. What's the plan to avoid the drift?

rexagod committed:
MON-4386,MON-4387: Introduce telemetry collection profile

This profile only caters to metrics that telemetry rules rely upon. The
patch adds a jsonnet script to broadly help with the following:
* generate the telemetry whitelist with added metadata to better track
  each rule, and,
* generate the telemetry service monitors for each component that we
  own, based on the entries teams add.

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

rexagod commented:
Please note that the fixup commit includes all generated data, the one before that has the actual changes. LMK if there's anything I can do here to make reviewing easier.

rexagod commented:
- [ ] Add test that queries ALL telemetry rules when the profile is applied, to make sure the e2e behavior is preserved.

rexagod committed:
MON-4386,MON-4387: Fix malformed generated monitor names

`rstripChars` was not just dropping the suffix, but anything from the
end of the string to the point where none of the characters specified
matched anymore. This fixes that.

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

rexagod committed:
MON-4386,MON-4387: Add test for telemetry profile

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

rexagod commented on manifests/0000_50_cluster-monitoring-operator_04-config.yaml:
Comparing the rules here with the initial commit in this PR (which had unchanged whitelist from main at that time), there is no difference between the rules exposed (only their metadata).

rexagod commented:
Rebase errors are in the generated docs, and shouldn't block reviews. I'll add the newer telemetry metrics in the jsonnet file and generate once this is good to go.

simonpasquier reviewed (commented):
I commented already on the need of implementing the Alertmanager's minimal profile before but it applies to all components that send metrics to telemetry.

juzhao commented on test/e2e/telemeter_test.go:
```diff
@@ -18,13 +18,34 @@ import (
 )
 
+// Some metrics may only be conditionally present depending on the environment, or the configuration.
+// For such metics, we skip the presence check.
```
```suggestion
// For such metrics, we skip the presence check.
```

rexagod commented on jsonnet/utils/telemetry-allowlist-and-monitors-entries.libsonnet:
I'll create a seperate PR.

rexagod commented:
I've added the pending TODOs in the PR description.

rexagod committed:
OCPBUGS-74524: Remove MetricsCollectionProfiles feature-gate

Blocks #2694

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

rexagod committed:
MON-4386,MON-4387: Drop extraneous tabs

juzhao commented:
/test okd-scos-images

rexagod commented on test/e2e/telemeter_test.go:
```diff
@@ -18,13 +18,34 @@ import (
+// For such metrics, we skip the presence check.
+var ignoredMetrics = map[string]struct{}{
+	"kube_pod_restart_policy":                                       {}, // Disabled through --metric-denylist.
+	"kubelet_volume_stats_used_bytes":                               {},
```
Every metric ignored here is used in the whitelist, however, they are not exposed OOTB in a fresh cluster without additional configuration.

Please let me know if you find otherwise.

simonpasquier commented:
I believe that this PR can be closed?

rexagod commented:
/close

I'll open a new PR to salvage the whitelist generation from here.
